### ISSUE TITLE
Add a utility function to copy a responses and modify its headers

### ISSUE THREAD

**Issue by philipwalton** [GoogleChrome/workbox#2129]
Forking this feature request from #2023: copying a response and modifying its headers is probably a fairly common thing to want to do, but actually implementing it in a performant and cross-browser way is non-trivial.

I think we could provide a utility function to do this in `workbox-core`. Sample usage might look something like this:

```js
const newResponse = await copyResponseAndModifyHeaders(oldResponse, (headers) => {
  // Any modifications to the `headers` object get copied to the new
  headers.append('X-Cache-Hit', '1');
})
```

In this example, the `X-Cache-Hit` header is added so analytics on the main page can know that this response came from the cache and not the network.

I don't 100% understand they history behind when a similar copy technique was used for `cleanRedirect()`, but #2023 shows that doing it right, and cross-browser is tricker than it might initially seem.

**Comment by nhoizey** [comment]
I never had this need yet, but it looks useful.

What about this:

```javascript
const newResponse = await copyResponseAndModifyBody(oldResponse, (body) => {
  // Any modifications to the `headers` object get copied to the new
  body.replace('src="https://my-domain.com', 'src="https://images-cdn.com');
})
```

Just a probably dumb example, but is the idea useful?

**Comment by philipwalton** [comment]
Hmmm, in my suggestion the copied response would share the body stream of the original response, so modifying it wouldn't be possible with a `replace()` function like you have there.

I'm not sure if there'd be an easy way to make body modification work in both the streaming and non-streaming case, or if it's worth doing that considering Edge is moving to Chromium in the near future, and then all SW supporting browsers should also support streams (I believe).

**Comment by nhoizey** [comment]
You're right, I didn't think about streams, I'm not yet used to it.

**Comment by philipwalton** [comment]
Closed via #2144
